### PR TITLE
[DEVHAS-101] Application Service Binding Controller

### PR DESCRIPTION
**Pull request by maysunfaisal**
This PR implements:
- controller for ApplicationSnapshotEnvironmentBinding
- ApplicationSnapshotEnvironmentBinding tests
- Updates Readme

#### To test it out:
- Create app `k create -f application/app.yaml`
- Create 2 components `component-sample` & `component-sample2` as per the binding spec `k create -f component/component-basic.yaml`
- Create snapshot `k apply -f snapshot/snapshot.yaml`
- Create binding `k apply -f binding/binding.yaml`

After the binding is created, describe the binding and check the status to see if the GitOps repo refresh was successful and check the GitOps repo for the overlays dir on both the components

Happy reviewing 😄

### CODE REVIEW

File: controllers/applicationsnapshotenvironmentbinding_controller.go [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+		return ctrl.Result{}, fmt.Errorf("unable to create temp directory for gitops resources due to error: %v", err)
+	}
+
+	// Get the Snapshot CR
+	appSnapshot := appstudioshared.ApplicationSnapshot{}
+	err = r.Get(ctx, types.NamespacedName{Name: snapshotName, Namespace: appSnapshotEnvBinding.Namespace}, &appSnapshot)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("unable to get the Application Snapshot %s %v", snapshotName, req.NamespacedName))
+		r.SetCreateConditionAndUpdateCR(ctx, &appSnapshotEnvBinding, err)
+		return ctrl.Result{}, nil
```
Comment (johnmcollier) [inline controllers/applicationsnapshotenvironmentbinding_controller.go:101]: We should requeue here (along with any other instances where we're not requeueing in the controller)

Returning err instead of nil will trigger the requeue.

File: controllers/applicationsnapshotenvironmentbinding_controller.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+			err := fmt.Errorf("application snapshot %s did not reference component %s", snapshotName, componentName)
+			log.Error(err, "")
+			r.SetCreateConditionAndUpdateCR(ctx, &appSnapshotEnvBinding, err)
+			return ctrl.Result{}, nil
+		}
+
+		gitopsStatus := hasComponent.Status.GitOps
+		// Get the information about the gitops repository from the Component resource
+		gitOpsURL = gitopsStatus.RepositoryURL
+		if gitOpsURL == "" {
```
Comment (yangcao77) [inline controllers/applicationsnapshotenvironmentbinding_controller.go:153]: can we add unit tests on  empty `gitOpsURL`, empty `context`, and non-empty `branch`? for test coverage

File: config/samples/snapshot/snapshot.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# API type definition is in github.com/redhat-appstudio/managed-gitops
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: ApplicationSnapshot
+metadata:
+  name: applicationsnapshot-sample
+spec:
+  application: application-sample
+  displayName: my-snapshot
+  displayDescription: my first snapshot
+  components:
```
Comment (johnmcollier) [inline config/samples/snapshot/snapshot.yaml:10]: What if the same component is referenced more than once here? Is that valid / Do we check for that and validate?

File: controllers/applicationsnapshotenvironmentbinding_controller.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+			r.SetCreateConditionAndUpdateCR(ctx, &appSnapshotEnvBinding, err)
+			return ctrl.Result{}, nil
+		}
+
+		var clone bool
+		var imageName string
+
+		for _, snapshotComponent := range appSnapshot.Spec.Components {
+			if snapshotComponent.Name == componentName {
+				imageName = snapshotComponent.ContainerImage
```
Comment (johnmcollier) [inline controllers/applicationsnapshotenvironmentbinding_controller.go:139]: Should we `break` here rather than continuing to loop?

File: controllers/applicationsnapshotenvironmentbinding_controller.go [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+		gitopsStatus := hasComponent.Status.GitOps
+		// Get the information about the gitops repository from the Component resource
+		gitOpsURL = gitopsStatus.RepositoryURL
+		if gitOpsURL == "" {
+			err := fmt.Errorf("unable to create gitops resource, GitOps Repository not set on component status")
+			log.Error(err, "")
+			r.SetCreateConditionAndUpdateCR(ctx, &appSnapshotEnvBinding, err)
+			return ctrl.Result{}, nil
+		}
+		if gitopsStatus.Branch != "" {
```
Comment (johnmcollier) [inline controllers/applicationsnapshotenvironmentbinding_controller.go:159]: Should we have some common utils function to handle these default values for Branch/Context? IIRC, we have similar logic in either the Application/Component controller.

File: controllers/applicationsnapshotenvironmentbinding_controller.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+		} else {
+			gitOpsBranch = "main"
+		}
+		if gitopsStatus.Context != "" {
+			gitOpsContext = gitopsStatus.Context
+		} else {
+			gitOpsContext = "/"
+		}
+
+		if remoteURL == "" {
```
Comment (yangcao77) [inline controllers/applicationsnapshotenvironmentbinding_controller.go:170]: I'm wondering what will happen if the `remoteURL` is being set by previous component in the loop?

File: controllers/applicationsnapshotenvironmentbinding_controller.go [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+			// Construct the remote URL for the gitops repository
+			parsedURL, err := url.Parse(gitOpsURL)
```
Comment (johnmcollier) [inline controllers/applicationsnapshotenvironmentbinding_controller.go:172]: Similar as above, same logic that's in the Component controller. Should we extract out to a utils function?

File: gitops/generate.go [hunk 8]
```diff
@@ -85,6 +87,31 @@ func Generate(fs afero.Afero, gitOpsFolder string, outputFolder string, componen
 	return GenerateParentKustomize(fs, gitOpsFolder, commonStorage)
 }
 
+// GenerateOverlays generates the overlays dir during the binding reconcile
```
Comment (johnmcollier) [inline gitops/generate.go:90]: Nit: Don't refer to the binding reconcile here, maybe just something like "GenerateOverlays generates the overlays dir from a given BindingComponent"?

File: gitops/gitops.go [hunk 9]
```diff
@@ -76,7 +77,52 @@ func GenerateAndPush(outputPath string, remote string, component appstudiov1alph
...
+	if out, err := e.Execute(repoPath, "git", "add", "."); err != nil {
+		return fmt.Errorf("failed to add files for component %q to repository in %q %q: %s", componentName, repoPath, string(out), err)
+	}
+
+	// See if any files changed, and if so, commit and push them up to the repository
+	if out, err := e.Execute(repoPath, "git", "--no-pager", "diff", "--cached"); err != nil {
+		return fmt.Errorf("failed to check git diff in repository %q %q: %s", repoPath, string(out), err)
+	} else if string(out) != "" {
+		// Commit the changes and push
+		if out, err := e.Execute(repoPath, "git", "commit", "-m", fmt.Sprintf("Generate GitOps overlays resources for component %s and environment %s", componentName, environmentName)); err != nil {
```
Comment (johnmcollier) [inline gitops/gitops.go:125]:
```suggestion
		if out, err := e.Execute(repoPath, "git", "commit", "-m", fmt.Sprintf("Generate %s environment overlays for component %s", environmentName, componentName)); err != nil {
```""

File: gitops/gitops.go [hunk 10]
```diff
@@ -76,7 +77,52 @@ func GenerateAndPush(outputPath string, remote string, component appstudiov1alph
...
+
+	return nil
+}
+
+// GenerateOverlaysAndPush generates the overlays kustomize from App Env Snapshot Binding Spec
+func GenerateOverlaysAndPush(outputPath string, clone bool, remote string, component appstudioshared.BindingComponent, applicationName, environmentName, imageName, namespace string, e Executor, appFs afero.Afero, branch string, context string, componentGeneratedResources map[string][]string) error {
+	componentName := component.Name
+	repoPath := filepath.Join(outputPath, applicationName)
+
+	if clone {
```
Comment (johnmcollier) [inline gitops/gitops.go:96]: What's the `clone` flag for? Wouldn't we always need to clone to get the most up-to-date GitOps repository?

File: controllers/applicationsnapshotenvironmentbinding_controller.go [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+			}
+			parsedURL.User = url.User(r.GitToken)
+			remoteURL = parsedURL.String()
+			clone = true
+		}
+
+		isStatusUpdated := false
+		for _, bindingStatusComponent := range appSnapshotEnvBinding.Status.Components {
+			if bindingStatusComponent.Name == componentName {
+				isStatusUpdated = true
```
Comment (yangcao77) [inline controllers/applicationsnapshotenvironmentbinding_controller.go:186]: break

File: controllers/applicationsnapshotenvironmentbinding_controller.go [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+	r.SetCreateConditionAndUpdateCR(ctx, &appSnapshotEnvBinding, nil)
+
+	log.Info(fmt.Sprintf("Finished reconcile loop for %v", req.NamespacedName))
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ApplicationSnapshotEnvironmentBindingReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		// Uncomment the following line adding a pointer to an instance of the controlled resource as an argument
```
Comment (yangcao77) [inline controllers/applicationsnapshotenvironmentbinding_controller.go:238]: cleanup the comment?

File: gitops/gitops.go [hunk 13]
```diff
@@ -76,7 +77,52 @@ func GenerateAndPush(outputPath string, remote string, component appstudiov1alph
...
+			if out, err := e.Execute(repoPath, "git", "checkout", "-b", branch); err != nil {
+				return fmt.Errorf("failed to checkout branch %q in %q %q: %s", branch, repoPath, string(out), err)
+			}
+		}
+	}
+
+	// Generate the gitops resources and update the parent kustomize yaml file
+	gitopsFolder := filepath.Join(repoPath, context)
+	componentEnvOverlaysPath := filepath.Join(gitopsFolder, "components", componentName, "overlays", environmentName)
+	if err := GenerateOverlays(appFs, gitopsFolder, componentEnvOverlaysPath, component, imageName, namespace, componentGeneratedResources); err != nil {
```
Comment (yangcao77) [inline gitops/gitops.go:112]: I think you need to clean up the folder before calling `GenerateOverlays`.
Have you tried out what will happen if the overlays folder & files already exist?
IIRC, `fs.Create` will return error if the file exists. https://github.com/redhat-appstudio/application-service/blob/daa9c484120c4485af504b17059f86d518c3b887/gitops/yaml/resources.go#L58

File: config/samples/snapshot/snapshot.yaml (hunk 3 above)
Comment (maysunfaisal) [reply config/samples/snapshot/snapshot.yaml:10]: not sure how much more validation we want to do, I just did some basic sanity validation

I feel the gitops webhook should be validating these semantics or whoever is the primary controller. But your point stands, it is a valid case and i have not done it in this PR

File: gitops/gitops.go (hunk 10 above)
Comment (maysunfaisal) [reply gitops/gitops.go:96]: so on a new reconcile call it will always be true and will be cloned and checked out

but since we in the controller we're looping the binding component(s) and generate fn is called per component, we only clone for the first component and reuse the same cloned files for the remaining components

File: gitops/gitops.go (hunk 13 above)
Comment (maysunfaisal) [reply gitops/gitops.go:112]: Need to test this out but will defer it to a new issue potentially if it causes to break

File: config/kcp/apiexport_has.yaml [hunk 14]
```diff
@@ -7,6 +7,6 @@ metadata:
   name: has
 spec:
   latestResourceSchemas:
```
[the remark this answers is not in the document]
Comment (maysunfaisal) [reply config/kcp/apiexport_has.yaml:9]: But I wonder if this is actually HAS' responsibility? Here are my arguments:
- All the team's controllers would be eventually working on kcp in unison, so would it not be the primary controller's responsibility to generate `APIResourceSchema` for their respective APIs? Same way I would assume build-service wouldnt be generating an `APIResourceSchema` for `Component` even when they have a controller for `Component`
- When we're using HAS controllers in dev/staging; we need to explicitly install the binding, snapshot, etc CRDs; otherwise our controller will error out. So I am wondering if this is any different to how the aforementioned teams like build-service, managed-gitops, etc dont really do install `Component` CRDs even if they have a controller for it 

🤔

File: controllers/applicationsnapshotenvironmentbinding_controller_conditions.go [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (r *ApplicationSnapshotEnvironmentBindingReconciler) SetCreateConditionAndUpdateCR(ctx context.Context, appSnapshotEnvBinding *appstudioshared.ApplicationSnapshotEnvironmentBinding, createError error) {
+	log := r.Log.WithValues("ApplicationSnapshotEnvironmentBinding", appSnapshotEnvBinding.Name)
+
+	if createError == nil {
+		meta.SetStatusCondition(&appSnapshotEnvBinding.Status.GitOpsRepoConditions, metav1.Condition{
+			Type:    "Sync",
```
[the remark this answers is not in the document]
Comment (johnmcollier) [reply controllers/applicationsnapshotenvironmentbinding_controller_conditions.go:32]: `GitOpsResourcesGenerated`

File: config/kcp/apiexport_has.yaml (hunk 14 above)
[the remark this answers is not in the document]
Comment (johnmcollier) [reply config/kcp/apiexport_has.yaml:9]: You're correct that it's not our responsibility, but I was making a note just that this will complicate testing on KCP until `ThingPermissionClaim` support is in KCP, and the APIResourceSchemas / APIExports for the Snapshot & Binding CRDs are generated by GitOps.

[the remark this answers is not in the document]
Comment (johnmcollier) [reply config/kcp/apiexport_has.yaml:9]: Until then, we'll need to manually generate the APIResourceSchemas and add it to our APIExports before testing.

[the remark this answers is not in the document]
Comment (maysunfaisal) [reply config/kcp/apiexport_has.yaml:9]: ok, do we just copy paste the CR spec in the field there for these new CRs? 🤔

File: controllers/applicationsnapshotenvironmentbinding_controller_conditions.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (maysunfaisal) [reply controllers/applicationsnapshotenvironmentbinding_controller_conditions.go:32]: I'll change to `GitOpsResourcesGenerated`

File: controllers/applicationsnapshotenvironmentbinding_controller_conditions.go [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+import (
+	"context"
+	"fmt"
+
+	appstudioshared "github.com/redhat-appstudio/managed-gitops/appstudio-shared/apis/appstudio.redhat.com/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (r *ApplicationSnapshotEnvironmentBindingReconciler) SetCreateConditionAndUpdateCR(ctx context.Context, appSnapshotEnvBinding *appstudioshared.ApplicationSnapshotEnvironmentBinding, createError error) {
```
[the remark this answers is not in the document]
Comment (maysunfaisal) [reply controllers/applicationsnapshotenvironmentbinding_controller_conditions.go:27]: Im updating to `SetConditionAndUpdateCR` since I am changing to `GitOpsResourcesGenerated`

File: controllers/applicationsnapshotenvironmentbinding_controller.go (hunk 6 above)
Comment (maysunfaisal) [reply controllers/applicationsnapshotenvironmentbinding_controller.go]: this was tied to the clone logic, i have removed it in my recent commit

File: Makefile [hunk 17]
```diff
@@ -275,7 +275,8 @@ apply-crds:
 	kubectl apply \
 	-f $(APPLICATIONS_CRD) \
 	-f $(COMPONENT_DETECTION_QUERIES_CRD) \
-	-f $(COMPONENT_CRD)
+	-f $(COMPONENT_CRD) \
+	-f https://raw.githubusercontent.com/redhat-appstudio/managed-gitops/main/appstudio-shared/manifests/appstudio-shared-customresourcedefinitions.yaml
```
Comment (johnmcollier) [inline Makefile:279]: Should the `install` makefile target be updated to apply the shared CRDs? Then we can cut out the section in the readme about installing the crd.

Comment (maysunfaisal) [reply Makefile]: updated

File: pkg/util/util.go [hunk 18]
```diff
@@ -49,6 +50,36 @@ func IsExist(path string) (bool, error) {
 	return false, err
 }
 
+// ProcessGitOpsStatus processes the GitOps status and returns the remote url, branch, context and the error
+func ProcessGitOpsStatus(gitopsStatus appstudiov1alpha1.GitOpsStatus, gitToken string) (string, string, string, error) {
```
[the remark this answers is not in the document]
Comment (johnmcollier) [reply pkg/util/util.go:54]: I guess what I'm getting at is, wouldn't be it be useful for the GitOpsStatus to say that the branch is `main` or that the context is `/`?

[the remark this answers is not in the document]
Comment (maysunfaisal) [reply pkg/util/util.go:54]: discussed in the meeting, we will keep it as it is and open an issue to track this as we would need to up date some of the logic. For instance, we cannot assume a repo's default branch is `main`

**Review by johnmcollier — APPROVED** [review]
